### PR TITLE
Add dump_checkpoint_fn interface into AA to control dumping checkpoints

### PR DESCRIPTION
**Pull request by kshpv**
This PR extends current Accuracy Aware interface and introduce a new user defined optional function - `dump_checkpoint_fn(model, compression_controller, accuracy_aware_runner, save_dir)`. 
    An (optional) function that allows a user to control the model's checkpoint saving process.
    Training loop will call this function instead own dump_checkpoint function

### CODE REVIEW

File: nncf/common/accuracy_aware_training/runner.py [hunk 1]
```diff
@@ -235,11 +236,12 @@ def __init__(self, accuracy_aware_params: Dict[str, object], verbose=True,
         self.cumulative_epoch_count = 0
         self.best_val_metric_value = 0
 
-    def initialize_training_loop_fns(self, train_epoch_fn, validate_fn, configure_optimizers_fn,
+    def initialize_training_loop_fns(self, train_epoch_fn, validate_fn, configure_optimizers_fn, dump_checkpoint_fn,
```
Comment (alexsu52) [inline nncf/common/accuracy_aware_training/runner.py:239]: Could you provide a signature for `dump_checkpoint_fn`? I think the signature should be common for PyTorch and TensorFlow backends. For example, `dump_checkpoint_fn(model, compression_controller, accuracy_aware_runner, save_dir)` 

Don't forget to update `Usage.md` with your changes.

Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py:239]: I am suggesting this signature - `dumping_fn(self, model, save_dir, accuracy_aware_metainfo)`, where `accuracy_aware_metainfo` is all accuracy_aware metainfo including compression_controller state

File: nncf/common/accuracy_aware_training/runner.py [hunk 2]
```diff
@@ -104,17 +104,18 @@ def calculate_minimal_tolerable_accuracy(self, uncompressed_model_accuracy):
...
         Register the user-supplied functions to be used to control the training process.
 
         :param train_epoch_fn: a method to fine-tune the model for a single epoch
         (to be called inside the `train_epoch` of the TrainingRunner).
-        :param validate: a method to evaluate the model on the validation dataset
+        :param validate_fn: a method to evaluate the model on the validation dataset
         (to be called inside the `train_epoch` of the TrainingRunner).
         :param configure_optimizers_fn: a method to instantiate an optimizer and a learning
         rate scheduler (to be called inside the `configure_optimizers` of the TrainingRunner).
+        :param dump_checkpoint_fn: a method to dump checkpoint with all necessary information
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py:118]: Done

File: nncf/common/accuracy_aware_training/training_loop.py [hunk 3]
```diff
@@ -52,8 +52,9 @@ def run(self, model: ModelType, train_epoch_fn, validate_fn,
         :param model: The model instance before fine-tuning
         :param train_epoch_fn: a method to fine-tune the model for a single epoch
         (to be called inside the `train_epoch` of the TrainingRunner)
-        :param validate: a method to evaluate the model on the validation dataset
+        :param validate_fn: a method to evaluate the model on the validation dataset
         (to be called inside the `train_epoch` of the TrainingRunner)
+        :param dump_checkpoint_fn: a method to dump a checkpoint with all necessary information
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/accuracy_aware_training/training_loop.py]: Done

**Comment by kshpv** [comment]
Please, take a look @vshampor @alexsu52 @AlexanderDokuchaev

**Review by AlexanderDokuchaev — APPROVED** [review]

**Review by vshampor — APPROVED** [review]

File: nncf/common/accuracy_aware_training/runner.py (hunk 1 above)
Comment (alexsu52) [reply nncf/common/accuracy_aware_training/runner.py]: If I right understand, you add to a new entity to the public API. I mean `accuracy_aware_metainfo` which AccuracyAwareTrainingRunner provides. At least, we should document keys in the accuracy_aware_metainfo dict.

Could you add test for TensorFlow AA Impl to save checkpoint?

File: nncf/torch/accuracy_aware_training/runner.py [hunk 4]
```diff
@@ -133,9 +135,7 @@ def dump_statistics(self, model, compression_controller):
                     self.add_tensorboard_scalar('compression/statistics/{0}'.format(key),
                                                 value, self.cumulative_epoch_count)
 
-
-    def dump_checkpoint(self, model, compression_controller):
-        checkpoint_path = osp.join(self._checkpoint_save_dir, 'acc_aware_checkpoint_last.pth')
+    def _collect_checkpoint(self, model, compression_controller):
```
Comment (andrey-churkin) [inline nncf/torch/accuracy_aware_training/runner.py:138]: I think this method is unnecessary.

**Comment by andrey-churkin** [comment]
@kshpv Could you please describe the reason for these changes?

File: nncf/torch/accuracy_aware_training/runner.py (hunk 4 above)
Comment (kshpv) [reply nncf/torch/accuracy_aware_training/runner.py:138]: I just added to improve code's readability

File: nncf/common/accuracy_aware_training/runner.py (hunk 1 above)
Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py]: > @kshpv Could you please describe the reason for these changes?

Was added

**Comment by kshpv** [comment]
> @kshpv Could you please describe the reason for these changes?

Was added
